src/core/dynamic-user.c

```diff
@@ -0,0 +1,786 @@
+         *
+         * b) a datagram containing a UID stored, but no lock fd attached to it. In this case there was already a
+         *    statically assigned UID by the same name, which we are reusing.
+         *
+         * c) a datagram containing a UID stored, and a lock fd is attached to it. In this case we allocated a dynamic
+         *    UID and locked it in the file system, using the lock fd.
+         *
+         * As PID 1 and various children might access the socket pair simultaneously, and pop the datagram or push it
+         * back in any time, we also maintain a lock on the socket pair. Note one peculiarity regarding locking here:
+         * the UID lock on disk is protected via a BSD file lock (i.e. an fd-bound lock), so that the lock is kept in
+         * place as long as there's a reference to the fd open. The lock on the storage socket pair however is a POSIX
+         * file lock (i.e. a process-bound lock), as all users share the same fd of this (after all it is anonymous,
+         * nobody else could get any access to it except via our own fd) and we want to synchronize access between all
+         * processes that have access to it. */
+
+        d = hashmap_get(m->dynamic_users, name);
+        if (d) {
+                /* We already have a structure for the dynamic user, let's increase the ref count and reuse it */
+                d->n_ref++;
+                *ret = d;
+                return 0;
+        }
+
+        if (!valid_user_group_name_or_id(name))
+                return -EINVAL;
+
+        if (socketpair(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0, storage_socket) < 0)
+                return -errno;
+
+        r = dynamic_user_add(m, name, storage_socket, &d);
+        if (r < 0)
+                return r;
+
+        storage_socket[0] = storage_socket[1] = -1;
+
+        if (ret) {
+                d->n_ref++;
+                *ret = d;
+        }
+
+        return 1;
+}
+
+static int pick_uid(const char *name, uid_t *ret_uid) {
+
+        static const uint8_t hash_key[] = {
+                0x37, 0x53, 0x7e, 0x31, 0xcf, 0xce, 0x48, 0xf5,
+                0x8a, 0xbb, 0x39, 0x57, 0x8d, 0xd9, 0xec, 0x59
+        };
+
+        unsigned n_tries = 100;
+        uid_t candidate;
+        int r;
+
+        /* A static user by this name does not exist yet. Let's find a free ID then, and use that. We start with a UID
+         * generated as hash from the user name. */
+        candidate = UID_CLAMP_INTO_RANGE(siphash24(name, strlen(name), hash_key));
+
+        (void) mkdir("/run/systemd/dynamic-uid", 0755);
+
+        for (;;) {
+                char lock_path[strlen("/run/systemd/dynamic-uid/") + DECIMAL_STR_MAX(uid_t) + 1];
+                _cleanup_close_ int lock_fd = -1;
+                ssize_t l;
+
+                if (--n_tries <= 0) /* Give up retrying eventually */
+                        return -EBUSY;
+
+                if (candidate < UID_PICK_MIN || candidate > UID_PICK_MAX)
```

Comment:
That is true indeed. Note however, that this code is somewhat inspired by the userns code in nspawn (where we allocate a UID range randomly for a container when we start it). There we initialize the candidate value from the stat() data of the container's root dir, and then validate it before using it, only falling back to randomized UID range allocation on conflict. Now, the currently used algorithm in this PR for the allocation of the per-service dynamic UIDs doesn't do something similar yet, however it's my plan to add that shortly, and hence I figured the logic could be kept in sync with the nspawn uid allocation loop even if the checks are redundant currently.

Specifically I'd like to add StateDirectory= and CacheDirectory= as service settings, that may be used to declare directories in /var/lib and /var/cache as owned by a service. Listed dirs would be created/chowned when the services is started, and, most importanly would -- via the stat() data - be used as start value for the UID for the dynamic UID algorithm.

(StateDirectory= and CacheDirectory= are a bit more complicated in the end than described in the paragraph above. Specifically, when used in conjunction with DynamicUser=1 they'll probably not bind mount /var/lib/foo → /var/lib/foo into the namspace of the container, but instead /var/lib/dynamic/foo → /var/lib/foo, and /var/lib/dynamic would be be a directory with an access mode of 700 or less. This is necessary so that services that might get the same UID assigned as a previously run service can't get access to the previously created state dir... This is very similar to how /var/lib/machines/ has an access mode of 700 or less, so that host programs can't take advantage of UIDs assigned different in containers attached to files in its directory tree. But anyway, that's probably something to discuss in the final PR for that feature, I just wanted to mention this here as background).